File: 3606-coupon-code-validator/3606-coupon-code-validator.py
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        res = []
        lines = {"electronics", "grocery", "pharmacy", "restaurant"}
        for i in range(len(code)):
            if all(c.isalnum() or c == '_' for c in code[i]) and bool(code[i]) and businessLine[i] in lines and isActive[i]:
                res.append((code[i], businessLine[i]))
        res = sorted(res, key=lambda x: (x[1], x[0]))
        return [x[0] for x in res]

File: 3606-coupon-code-validator/3606-coupon-code-validator.cpp
class Solution {
public:
    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        vector<string> res;
        vector<pair<string, string>> resPairs;
        unordered_set<string> lines = {"electronics", "grocery", "pharmacy", "restaurant"};
        for(int i = 0; i < code.size(); i++)
        {
            if(nonSpecial(code[i]) && lines.count(businessLine[i]) && isActive[i])
            {
                resPairs.push_back({businessLine[i], code[i]});
            }
        }
        sort(resPairs.begin(), resPairs.end());
        for(int i = 0; i < resPairs.size(); i++)
        {
            res.push_back(resPairs[i].second);
        }
        return res;
    }
private:
    bool nonSpecial(string& s)
    {
        if(s.empty())
        {
            return false;
        }
        for(int i = 0; i < s.length(); i++)
        {
            if(s[i] != '_' && !(s[i] >= 'a' && s[i] <= 'z') && !(s[i] >= 'A' && s[i] <= 'Z') && !(s[i] >= '0' && s[i] <= '9'))
            {
                return false;
            }
        }
        return true;
    }
};